src/test/java/io/vertx/core/impl/launcher/commands/WatcherTest.java

```diff
@@ -0,0 +1,176 @@
+          if (next != null) {
+            next.handle(null);
+          }
+        },
+        null);
+  }
+
+  @After
+  public void close() {
+    watcher.close();
+  }
+
+  @Test
+  public void testFileAddition() throws IOException {
+    watcher.watch();
+
+    // Initial deployment
+    waitUntil(() -> deploy.get() == 1);
+
+    File file = new File(root, "foo.txt");
+    file.createNewFile();
+
+    // undeployment followed by redeployment
+    waitUntil(() -> undeploy.get() == 1 && deploy.get() == 2);
+  }
+
+  @Test
+  public void testWithANonMatchingFile() throws IOException, InterruptedException {
+    watcher.watch();
+
+    // Initial deployment
+    waitUntil(() -> deploy.get() == 1);
+
+    File file = new File(root, "foo.nope");
+    file.createNewFile();
+
+    Thread.sleep(2000);
+    assertThat(undeploy.get()).isEqualTo(0);
+    assertThat(deploy.get()).isEqualTo(1);
+  }
+
+  @Test
+  public void testFileModification() throws IOException, InterruptedException {
+    File file = new File(root, "foo.txt");
+    file.createNewFile();
+
+    watcher.watch();
+
+    // Initial deployment
+    waitUntil(() -> deploy.get() == 1);
+
+    // Wait until the file monitoring is set up (ugly, but I don't know any way to detect this).
+    Thread.sleep(2000);
+    // Simulate a 'touch'
+    file.setLastModified(System.currentTimeMillis());
+
+    // undeployment followed by redeployment
+    waitUntil(() -> undeploy.get() == 1 && deploy.get() == 2);
+  }
+
+  @Test
+  public void testFileDeletion() throws IOException, InterruptedException {
+    File file = new File(root, "foo.txt");
+    file.createNewFile();
+
+    watcher.watch();
+
+    // Initial deployment
+    waitUntil(() -> deploy.get() == 1);
+
+    // Wait until the file monitoring is set up (ugly, but I don't know any way to detect this).
+    Thread.sleep(2000);
+    file.delete();
+
+    // undeployment followed by redeployment
+    waitUntil(() -> undeploy.get() == 1 && deploy.get() == 2);
+  }
+
+  @Test
+  public void testFileAdditionInDirectory() throws IOException, InterruptedException {
+    watcher.watch();
+    // Wait until the file monitoring is set up (ugly, but I don't know any way to detect this).
+    Thread.sleep(2000);
+
+    // Initial deployment
+    waitUntil(() -> deploy.get() == 1);
+
+    File newDir = new File(root, "directory");
+    newDir.mkdir();
+    Thread.sleep(2000);
+    File file = new File(newDir, "foo.txt");
+    file.createNewFile();
+
+    // undeployment followed by redeployment
+    waitUntil(() -> undeploy.get() == 1 && deploy.get() == 2);
+  }
+
+  public static boolean deleteRecursive(File path) {
+    if (!path.exists()) {
+      return false;
+    }
+    boolean ret = true;
+    if (path.isDirectory()) {
+      File[] files = path.listFiles();
+      if (files != null) {
+        for (File f : files) {
+          ret = ret && deleteRecursive(f);
+        }
+      }
+    }
+    return ret && path.delete();
+  }
+
+}
```

Comment:
Also test addition and deletion of directory? In particular:

Create new directory
Create new file in directory.
Amend file in directory.